### PR TITLE
Base refactor

### PR DESCRIPTION
**Pull request by Fr3E**

### CODE REVIEW

File: ImagePicker/ImagePickerAssetModel.swift [hunk 1]
```diff
@@ -1,89 +1,67 @@
...
-        log("deinit: \(String(describing: self))")
-    }
-    
     var fetchResult: PHFetchResult<PHAsset>! {
         set { userDefinedFetchResult = newValue }
         get { return userDefinedFetchResult ?? defaultFetchResult }
     }
     
+    // Will be use for caching
+    var previousPreheatRect = CGRect.zero
```
Comment (radimhalfar) [inline ImagePicker/ImagePickerAssetModel.swift:14]: if the class is called model, the properties should be private and unmodifiable from outside

File: ImagePicker/ImagePickerAssetModel.swift [hunk 2]
```diff
@@ -1,89 +1,67 @@
...
             return log("asset model: update cache assets - collection view layout is not flow layout")
         }
         
         // The preheat window is twice the height of the visible rect.
         let visibleRect = CGRect(origin: collectionView.contentOffset, size: collectionView.bounds.size)
-        
         var preheatRect: CGRect
         
         switch layout.scrollDirection {
         case .vertical:
```
Comment (radimhalfar) [inline ImagePicker/ImagePickerAssetModel.swift]: both cases are pretty much the same - can you extract it to function and unify it?

File: ImagePicker/ImagePickerDataSource.swift [hunk 3]
```diff
@@ -57,31 +47,32 @@ final class ImagePickerDataSource : NSObject, UICollectionViewDataSource {
...
         }
     }
     
-
+    private func assetItems(for indexPath: IndexPath, collectionView: UICollectionView, cellsRegistrator: CellRegistrator) -> UICollectionViewCell {
+        let asset = assetsModel.fetchResult.object(at: indexPath.item)
+        let cellId = cellsRegistrator.cellIdentifier(forAsset: asset.mediaType) ?? cellsRegistrator.cellIdentifierForAssetItems
+        
+        guard let cell = collectionView.dequeueReusableCell(withReuseIdentifier: cellId, for: indexPath) as? ImagePickerAssetCell else {
+            fatalError("asset item cell must conform to \(ImagePickerAssetCell.self) protocol")
```
Comment (radimhalfar) [inline ImagePicker/ImagePickerDataSource.swift]: assertion might be better - as dependency we don't want to crash the app, we want to notify developer he did something wrong

File: ImagePicker/ImagePickerLayout.swift [hunk 4]
```diff
@@ -130,4 +92,32 @@ final class ImagePickerLayout {
...
+    private func sizeForCameraItem(_ collectionView: UICollectionView, layoutModel: LayoutModel, layout: UICollectionViewFlowLayout) -> CGSize {
+        // Lets keep this ratio so camera item is a nice rectangle
+        let traitCollection = collectionView.traitCollection
+        var ratio: CGFloat = 160/212
+        
+        // For iphone in landscape we need different ratio
+        if traitCollection.userInterfaceIdiom == .phone {
+            switch (traitCollection.horizontalSizeClass, traitCollection.verticalSizeClass) {
+            // Iphones in landscape
+            case (.unspecified, .compact):
```
Comment (radimhalfar) [inline ImagePicker/ImagePickerLayout.swift]: put fallthrough cases into one

**Review by radimhalfar — CHANGES_REQUESTED** [review]

File: ImagePicker/ImagePickerAssetModel.swift (hunk 1 above)
Comment (Fr3E) [reply ImagePicker/ImagePickerAssetModel.swift:14]: Class renamed

File: ImagePicker/ImagePickerDataSource.swift (hunk 3 above)
Comment (Fr3E) [reply ImagePicker/ImagePickerDataSource.swift]: wont do

**Review by radimhalfar — APPROVED** [review]

**Pull request merged**